### DISCUSSION TITLE
How to have a separate format and css class select field when inserting image to richtext field?

### DISCUSSION THREAD

**Discussion by ordigital** [wagtail/wagtail#11852] [Q&A]
The purpose of filters (renditions) is to apply operations on the image file. The purpose of css classes is to apply styles – especially aligning. This is why it is necessary to have separate fields to select image format and image class for aligning. 

From what I see it's not possible and I have to do as many formats as many combination of format & image aligning as I need.  So for example if I want to have align to left, right and center and I want to have tree different size of image, I need 9 formats. Isn't that ridiculous since it would be enough to have two select fields (one for format and one for class) instead and each with just 3 options? 

I see that template for `select_format` is getting form from `ImageInsertionForm` but I cannot see a way to register custom form class to add css class select field. Any ideas how to get additional class select field?

**Comment by Ganesh-403** [comment]
@ordigital From what I can tell, **Wagtail does not currently support separate selectors for image rendition (filter) and CSS class in the rich text image chooser**.

The built-in image insertion workflow is based around **`Format`** objects, where each format combines:

- a rendition (`filter_spec`), and
- the CSS classes (`classnames`)

into a single selectable option. That's why the default options ("Full width", "Left-aligned", "Right-aligned") each define both the image size and the CSS class together. :contentReference[oaicite:0]{index=0}

So your understanding is correct: if you want 3 sizes × 3 alignments, the current API requires defining 9 formats.

I also couldn't find any documented extension point that allows replacing `ImageInsertionForm` with a custom form (for example, to add a second `<select>` for CSS classes) or registering an alternative image insertion dialog. The customization documented by Wagtail is around registering/unregistering `Format` objects and customizing how a format renders HTML, rather than customizing the chooser form itself. :contentReference[oaicite:1]{index=1}

If your goal is simply to avoid the combinatorial explosion of formats, a couple of alternatives might be worth considering:

- If the alignment can be inferred from the surrounding layout, define formats for **size only** and apply alignment through your templates or container CSS instead.
- If you need editor-controlled alignment, the current supported approach is unfortunately to define the required combinations as separate formats.

That said, I think your suggestion of having two independent selectors—one for the rendition/filter and one for the CSS class—would make sense from both a UX and maintenance perspective. It would reduce duplication while keeping image processing (renditions) separate from presentation (CSS classes).

If I've missed an internal hook for overriding the rich text image insertion form, I'd be interested to hear from the maintainers, but based on the current documentation I couldn't find one.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
